### DISCUSSION TITLE
Document the $args() function in /wheels/global/internal.cfm

### DISCUSSION THREAD

**Discussion by dhgassoc** [wheels-dev/wheels#1398] [Ideas]
I recently had to dive into the `startFormTag()` to fix a problem in my cfwheels app that uses htmx with csrf in adobe@2016.
The first line of startFormTag() is: 
`	$args(name = "startFormTag", args = arguments);
`
The comment header for the $args() is non-descriptive. 
Can someone explain and write a new comment header to explain what this function is used for and how to use args? and maybe some different examples of how to use it.  

```
**
 * Internal function.
 */
public void function $args(
	required struct args,
	required string name,
	string reserved = "",
	string combine = "",
	string required = ""
)  ... }
```

Also please also add this comment header to next release of cfwheels. 
Thanks Dan G.

**Comment by adamcameron** [comment]
Um... my initial reaction is that it's on a file called "internal.cfm" and it's prefixed with a `$` which is Wheels's weirdo way of indicating a method should be considered private.

So I suspect it's probably not documented as it's not intended to be called outside of Wheels's own code..?

I'd be questioning whether yer taking the right approach to whatever yer doing that involves you needing to know about that function.

I suspect the entrypoint for consuming devs is here: `https://guides.cfwheels.org/2.5.0/working-with-cfwheels/configuration-and-defaults#function-settings`?

That said, those docs are a bit scant too :-/

**Reply by dhgassoc** [reply]
I am a part time cfwheels developer (more of a hobby for me).  I enjoying **LEARNING new code** and I write code because it is fun. I am learning how to add HTMX to cfWheels.   I make apps for myself.  I sometimes have weeks (or even months) between the times I can spend coding.   My memory is not that good anymore.  Therefore I HAVE to write a few lines of comments for each line of code - to remind myself of what I have already debugged/coded.  

I was debugging an HTMX  with CSRF problem.  I was creating a custom startHtmxFormTag() that extends startFormTag(). To do that, I needed to understand / trace the code in startFormTag() in \wheels\view\forms.cfm.  

The first line to all these related form functions reads;
`	$args(name = "startFormTag", args = arguments);  `
But I digress, I will be posting more details about problems I had / How to create HTMX + CSRF + cfwheels + Adobe-2016 version later once my app is working .  For this discussion, it is not important.  

The crux of the matter is that there are only a few core-cfwheels contributors (for which I am very grateful for).  If we ever want more developers to even be able to contribute and understand the cfwheels core, more understanding / documentation is needed _about_ the core.  Most of the coldfusion cfwheels code I can follow and understand.   So when I came across this $args() function in \wheels\global\internal.cfm that does NOT return a value, and is doing something (to globals) that _I do not understand_, I ask questions -- so I can **Learn more** about how the core works.   If I have a question, someone else may also have a similar question.  

I understand that no one has time to write a cfwheels core concepts document - or create a diagram of how the core works.  And I don't expect core-developers to waste their time doing so.  Rereading the guides is very helpful in understanding parts of the core.  The flow of the guides are great.  
In the past, when I have spent time learning/reviewing the actual code in the core, it has actually helped improve my own coldfusion code.  

But in this specific instance I would think that it would be a very small quick request (for someone that is in-the-know) to document a small but frequently used function.  While it may seem trivial to you, it  may not be as trivial to some novice like me. 

P.S. I was able to get my startHtmxFormTag() working without understanding what $args() does -- So it is not really a big deal.

Thanks Dan.

**Reply by MvdO79** [reply]
I can relate to the above statement. I believe that @bpamiri is in the process of writing a Cfwheels book so it might be possible core insights are explained there.

**Comment by bpamiri** [comment]
The book project seems to be on the permanent back burner. I hope to get to it again at some point but life keeps getting in the way.

**Reply by MvdO79** [reply]
Totally understandable.

**Comment by bpamiri** [comment]
@dhgassoc I would definitely be interested in hearing about your experience with HTMX. I’ve been totally infactuated by the project since coming upon it. I’ve also read the book https://hypermedia.systems which describes the ideas behind its creation. Another interesting point is how the Django community has adopted HTMX.

**Reply by chapmandu** [reply]
I've doubled down on HTMX and it's changed my life. 

Here's a simple POC I put together a while ago. 

https://github.com/chapmandu/cfwheels-htmx-crud

**Reply by dhgassoc** [reply]
Chapmandu and Peter, thank you for your htmx code examples and plugin.  I have been already using themas a reference for building my htmx app.  and combining it with Peter Amiri's posting on the cfwheels blog. 

One minor problem with your POC:  New HTMX version [deprecated ](https://htmx.org/headers/hx-push/) "HX-Push";
		cfheader(name = "HX-Push", value = URLFor(route = "Examples"));
I changed to 
		cfheader(name = "HX-Push-Url", value = URLFor(route = "Examples"));
